pilot/pkg/proxy/envoy/v2/discovery.go

```diff
@@ -428,3 +425,60 @@ func (s *DiscoveryServer) handleUpdates(stopCh <-chan struct{}) {
 		}
 	}
 }
+
+func (s *DiscoveryServer) checkProxyNeedsFullPush(node *model.Proxy) bool {
+	full := false
+	s.proxyUpdatesMutex.Lock()
+	if _, ok := s.proxyUpdates[node.IPAddresses[0]]; ok {
+		full = true
+		delete(s.proxyUpdates, node.IPAddresses[0])
+	}
+	s.proxyUpdatesMutex.Unlock()
+	return full
+}
+
+func doSendPushes(stopCh <-chan struct{}, semaphore chan struct{}, queue *PushQueue, checkProxyNeedsFullPush func(node *model.Proxy) bool) {
+	// Signals that a push is done by reading from the semaphore, allowing another send on it.
+	doneFunc := func() {
+		<-semaphore
+	}
+	for {
+		select {
+		case <-stopCh:
+			return
+		default:
+			// We can send to it until it is full, then it will block until a pushes finishes and reads from it.
+			// This limits the number of pushes that can happen concurrently
+			semaphore <- struct{}{}
+
+			// Get the next proxy to push. This will block if there are no updates required.
+			client, info := queue.Remove()
+
+			go func() {
+				edsUpdates := info.edsUpdatedServices
+				proxyFull := info.full || checkProxyNeedsFullPush(client.modelNode)
+
+				if proxyFull {
+					// Setting this to nil will trigger a full push
+					edsUpdates = nil
+				}
+
+				select {
+				case client.pushChannel <- &XdsEvent{
```

Comment:
I  can not see the retry. What if this push failed?

Reply:
maybe blocked for an undefined time?

Reply:
Mainly the net io may hang. Since there is a timer, it looks good.

Reply:
if the send timeout happens then the push will be lost. maybe we need to retry there? wdyt?

Reply:
agree

Reply:
Is this done?